src/x509_cert_helper.cpp

```diff
@@ -0,0 +1,123 @@
+#include "mls/x509_cert_helper.h"
+#include "openssl/asn1.h"
+#include "openssl/bio.h"
+#include "openssl/err.h"
+#include "openssl/objects.h"
+#include "openssl/pem.h"
+#include "openssl/rand.h"
+#include <openssl/x509.h>
+#include <openssl/x509v3.h>
+
+namespace mls {
+
+///
+/// OpenSSL X509 Certificate
+///
+
+using EVP_PKEY_ptr = std::unique_ptr<EVP_PKEY, decltype(&::EVP_PKEY_free)>;
+using X509_ptr = std::unique_ptr<X509, decltype(&::X509_free)>;
+
+static std::runtime_error
+openssl_error()
+{
+	uint64_t code = ERR_get_error();
+	return std::runtime_error(ERR_error_string(code, nullptr));
+}
+
+
+struct OpenSSLCertificate : public X509Certificate {
```

Comment:
I wouldn't bother with the layer of indirection.  Just have one real certificate class; I'm agnostic as to whether you call it `OpenSSLCertificate` or just `Certificate`.